pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeSegmentPruner.java

```diff
@@ -0,0 +1,346 @@
+import org.apache.pinot.common.utils.request.RequestUtils;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.DateTimeFieldSpec;
+import org.apache.pinot.spi.data.DateTimeFormatSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The {@code TimeSegmentPruner} prunes segments based on their time column start & end time metadata stored in ZK. The pruner
+ * supports queries with filter (or nested filter) of EQUALITY and RANGE predicates.
+ */
+public class TimeSegmentPruner implements SegmentPruner {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TimeSegmentPruner.class);
+  private static final long MIN_START_TIME = 0;
+  private static final long MAX_END_TIME = Long.MAX_VALUE;
+  private static final Interval DEFAULT_INTERVAL = new Interval(MIN_START_TIME, MAX_END_TIME);
+  private static final char DELIMITER = '\0';
+  private static final String LEGACY_DELIMITER = "\t\t";
+  private static final char START_INCLUSIVE = '(';
+  private static final char END_INCLUSIVE = ')';
+  private static final String UNBOUNDED = "*";
+
+  private final String _tableNameWithType;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private final String _segmentZKMetadataPathPrefix;
+  private final String _timeColumn;
+  private final DateTimeFormatSpec _timeFormatSpec;
+
+  private volatile IntervalTree<String> _intervalTree;
+  private final Map<String, Interval> _intervalMap = new HashMap<>();
+
+  public TimeSegmentPruner(TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _tableNameWithType = tableConfig.getTableName();
+    _propertyStore = propertyStore;
+    _segmentZKMetadataPathPrefix = ZKMetadataProvider.constructPropertyStorePathForResource(_tableNameWithType) + "/";
+    _timeColumn = tableConfig.getValidationConfig().getTimeColumnName();
+    Preconditions
+        .checkNotNull(_timeColumn, "Time column must be configured in table config for table: %s", _tableNameWithType);
+
+    Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+    Preconditions.checkNotNull(schema, "Failed to find schema for table: %s", _tableNameWithType);
+    DateTimeFieldSpec dateTimeSpec = schema.getSpecForTimeColumn(_timeColumn);
+    Preconditions.checkNotNull(dateTimeSpec, "Field spec must be specified in schema for time column: %s of table: %s",
+        _timeColumn, _tableNameWithType);
+    _timeFormatSpec = new DateTimeFormatSpec(dateTimeSpec.getFormat());
+  }
+
+  @Override
+  public void init(ExternalView externalView, IdealState idealState, Set<String> onlineSegments) {
+    // Bulk load time info for all online segments
+    int numSegments = onlineSegments.size();
+    List<String> segments = new ArrayList<>(numSegments);
+    List<String> segmentZKMetadataPaths = new ArrayList<>(numSegments);
+    for (String segment : segments) {
```

Comment:
`segments` is always going to be empty here. Shouldn't this be `onlineSegments` ?
The call on line 104 will always result in index out of bounds.
Am I missing something here? Shouldn't the tests have caught this?

Reply:
Yes, this should be `onlineSegments`. The tests only covered cases that `onlineSegments` are empty. Let me send a pr to fix this and cover cases when `onlineSegments` are not empty.